crypto/rand/rand_lib.c

```diff
@@ -187,6 +193,71 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
     return ret;
 }
 
+/*
+ * Generate additional data that can be used for the drbg. The data does
+ * not need to contain entropy, but it's useful if it contains at least
+ * some bits that are unpredictable.
+ *
+ * Returns 0 on failure.
+ *
+ * On success it allocates a buffer at |*pout| and returns the length of
+ * the data. The buffer should get freed using OPENSSL_secure_clear_free().
+ */
+size_t rand_drbg_get_additional_data(unsigned char **pout, size_t max_len)
+{
+    RAND_POOL *pool;
+    CRYPTO_THREAD_ID thread_id;
+    size_t len;
+#ifdef OPENSSL_SYS_UNIX
+    pid_t pid;
+    struct timeval tv;
+#elif defined(OPENSSL_SYS_WIN32)
+    DWORD pid;
+    FILETIME ft;
+#endif
+#ifdef OPENSSL_CPUID_OBJ
+    unsigned long long tsc = 0;
+#endif
+
+    pool = RAND_POOL_new(0, 0, max_len);
+    if (pool == NULL)
+        return 0;
+
+#ifdef OPENSSL_SYS_UNIX
+    pid = getpid();
+    RAND_POOL_add(pool, (unsigned char *)&pid, sizeof(pid), 0);
+#elif defined(OPENSSL_SYS_WIN32)
+    pid = GetCurrentProcessId();
+    RAND_POOL_add(pool, (unsigned char *)&pid, sizeof(pid), 0);
+#endif
+
+    thread_id = CRYPTO_THREAD_get_current_id();
+    if (thread_id != 0)
+        RAND_POOL_add(pool, (unsigned char *)&thread_id, sizeof(thread_id), 0);
+
+#ifdef OPENSSL_CPUID_OBJ
+    tsc = OPENSSL_rdtsc();
+    if (tsc != 0)
+        RAND_POOL_add(pool, (unsigned char *)&tsc, sizeof(tsc), 0);
+#endif
+
+#ifdef OPENSSL_SYS_UNIX
+    if (gettimeofday(&tv, NULL) == 0)
+        RAND_POOL_add(pool, (unsigned char *)&tv, sizeof(tv), 0);
+#elif defined(OPENSSL_SYS_WIN32)
+    GetSystemTimeAsFileTime(&ft);
+    RAND_POOL_add(pool, (unsigned char *)&ft, sizeof(ft), 0);
+#endif
+
+    /* TODO: Use RDSEED? */
```

Comment:
Since we're generating new bytes, RDRAND might be appropriate even though it is feeding into the DRBG.

The other quick data source that comes to mind are the `clock_gettime(2)` timers:

```
#if _POSIX_C_SOURCE >= 199309L
    struct timespec ts;
    clockid_t cid;

# ifdef CLOCK_BOOTTIME
    cid = CLOCK_BOOTTIME;
# elif defined CLOCK_MONOTONIC
    cid = CLOCK_MONOTONIC;
# else
    cid = CLOCK_REALTIME;
# endif

    if (clock_gettime(cid, &ts) == 0)
       ...
#endif
```

Reply:
I've actually had clock_gettime(CLOCK_MONOTONIC) in a previous version of the patch but I've removed it in favor of just reading the TSC ourselves.

But clock_gettime(CLOCK_REALTIME) might be a useful replacement for gettimeofday(), it has a resolution ns instead of μs.